Source/Core/VideoBackends/Vulkan/ShaderCache.cpp

```diff
@@ -0,0 +1,1377 @@
+  {
+    // We don't insert null modules into the shader map since creation could succeed later on.
+    // e.g. we're generating bad code, but fix this in a later version, and for some reason
+    // the cache is not invalidated.
+    VkShaderModule module = Util::CreateShaderModule(value, value_size);
+    if (module == VK_NULL_HANDLE)
+      return;
+
+    m_shader_map.emplace(key, std::make_pair(module, false));
+  }
+
+  std::map<Uid, std::pair<VkShaderModule, bool>>& m_shader_map;
+};
+
+void ShaderCache::LoadShaderCaches()
+{
+  ShaderCacheReader<VertexShaderUid> vs_reader(m_vs_cache.shader_map);
+  m_vs_cache.disk_cache.OpenAndRead(GetDiskShaderCacheFileName(APIType::Vulkan, "VS", true, true),
+                                    vs_reader);
+
+  ShaderCacheReader<PixelShaderUid> ps_reader(m_ps_cache.shader_map);
+  m_ps_cache.disk_cache.OpenAndRead(GetDiskShaderCacheFileName(APIType::Vulkan, "PS", true, true),
+                                    ps_reader);
+
+  if (g_vulkan_context->SupportsGeometryShaders())
+  {
+    ShaderCacheReader<GeometryShaderUid> gs_reader(m_gs_cache.shader_map);
+    m_gs_cache.disk_cache.OpenAndRead(GetDiskShaderCacheFileName(APIType::Vulkan, "GS", true, true),
+                                      gs_reader);
+  }
+
+  ShaderCacheReader<UberShader::VertexShaderUid> uber_vs_reader(m_uber_vs_cache.shader_map);
+  m_uber_vs_cache.disk_cache.OpenAndRead(
+      GetDiskShaderCacheFileName(APIType::Vulkan, "UberVS", false, true), uber_vs_reader);
+  ShaderCacheReader<UberShader::PixelShaderUid> uber_ps_reader(m_uber_ps_cache.shader_map);
+  m_uber_ps_cache.disk_cache.OpenAndRead(
+      GetDiskShaderCacheFileName(APIType::Vulkan, "UberPS", false, true), uber_ps_reader);
+
+  SETSTAT(stats.numPixelShadersCreated, static_cast<int>(m_ps_cache.shader_map.size()));
+  SETSTAT(stats.numPixelShadersAlive, static_cast<int>(m_ps_cache.shader_map.size()));
+  SETSTAT(stats.numVertexShadersCreated, static_cast<int>(m_vs_cache.shader_map.size()));
+  SETSTAT(stats.numVertexShadersAlive, static_cast<int>(m_vs_cache.shader_map.size()));
+}
+
+template <typename T>
+static void DestroyShaderCache(T& cache)
+{
+  cache.disk_cache.Sync();
+  cache.disk_cache.Close();
+  for (const auto& it : cache.shader_map)
+  {
+    if (it.second.first != VK_NULL_HANDLE)
+      vkDestroyShaderModule(g_vulkan_context->GetDevice(), it.second.first, nullptr);
+  }
+  cache.shader_map.clear();
+}
+
+void ShaderCache::DestroyShaderCaches()
+{
+  DestroyShaderCache(m_vs_cache);
+  DestroyShaderCache(m_ps_cache);
+
+  if (g_vulkan_context->SupportsGeometryShaders())
+    DestroyShaderCache(m_gs_cache);
+
+  DestroyShaderCache(m_uber_vs_cache);
+  DestroyShaderCache(m_uber_ps_cache);
+
+  SETSTAT(stats.numPixelShadersCreated, 0);
+  SETSTAT(stats.numPixelShadersAlive, 0);
+  SETSTAT(stats.numVertexShadersCreated, 0);
+  SETSTAT(stats.numVertexShadersAlive, 0);
+}
+
+VkShaderModule ShaderCache::GetVertexShaderForUid(const VertexShaderUid& uid)
```

Comment:
Not sure if it'd be worth it but in the future we could merge these variants into one function with the following parameters:  map, uid, and a enum stating type.  We could then call the appropriate function based on that type (or just pass in a functor?).